c/datasetjson.c

```diff
@@ -1630,6 +1640,395 @@ void respondWithDataset(HttpResponse* response, char* absolutePath, int jsonMode
+      "ZSS1757I %s lockRC = %d, lockRSN = %x\n", __FUNCTION__ , lockRC , lockRSN );  
+
+
+    if ( lockRC != 0){
+
+
+
+        if (lockRC == 4)
+        {
+            zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_WARNING, 
+              "ZSS1773E %s isgenqTryExclusiveLock Unable to obtain exclusive access to Dataset or member\n"
+              , __FUNCTION__);
+            respondWithError(response,HTTP_STATUS_RESOURCE_CONFLICT,
+              "Unable to obtain exclusive access to Dataset or member");
+            return;
+        }
+        else        {
+            zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG, 
+              "ZSS1779E %s isgenqTryExclusiveLock failed, RC = %d, reason code = %08X, %04X\n" , __FUNCTION__, 
+              lockRC , lockRSN, lockRSN & 0xFFFF);
+            respondWithError(response,HTTP_STATUS_INTERNAL_SERVER_ERROR,
+              "isgenqTryExclusiveLock failed");
+            return;
+        }
+    }
+    else{
+        zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG, 
+              "ZSS1789I %s Lock is available\n", __FUNCTION__);
+    }
+        
+/* test lock end */
+
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+      "ZSS1793I %s GetExclusiveLock\n", __FUNCTION__);
+
+    lockRC = isgenqGetExclusiveLock( 
+        &MajorQNAME, 
+        &rname_parm, 
+        ISGENQ_SCOPE_SYSTEMS, 
+        &lockToken, 
+        &lockRSN);   
+    /* This function WAITs in case of contention.  This should not be a problem because it's preceeded by isgenqTryExclusiveLock */
+    /* the lock is ALWAYS released when the C program exits. */
+
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+      "ZSS1805I %s lockRC = %d, lockRSN = %x, PID = %d\n", __FUNCTION__ , lockRC , lockRSN, pid );      
+
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+      "ZSS1808I %s lockToken<", __FUNCTION__); 
+    for(i=0; i<sizeof(lockToken); i++)
+    {         
+        /*printf("%x",*(char *)&lockToken+i);*/
+        zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,"%x",*(lockToken.token+i));
+    }
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,">\n");    
+
+/* ----------------- */
+/* semaphore section */
+/* ----------------- */
+
+
+    int semaphoreID;
+
+
+      for(i=0; i < N_SEM_TABLE_ENTRIES; i++)
+      {  
+
+        if (sem_table_entry[i].sem_ID != 0                                  /* semaphore exists */      
+            && memcmp(sem_table_entry[i].dsn, dsn_member.dsn, 44) == 0          /* DSN matches */
+            && memcmp(sem_table_entry[i].mem, dsn_member.membername, 8) == 0   /* member matches */
+            ) /* already present */
+        {
+          zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+            "ZSS1945I %s dsn already in table in row %d with semaphoreID %D\n", __FUNCTION__, i, sem_table_entry[i].sem_ID );  
+          break;
+        }
+        if (sem_table_entry[i].sem_ID == 0) /* found empty slot in table */
```

Comment:
How do you serialize this semaphore table itself? What if are 10 requests checking this table?